translator/internaldata/oc_to_metrics.go

```diff
@@ -0,0 +1,315 @@
+	labels := getInternalLabels(ocMetric)
+	setDataPoints(ocMetric, metric, labels)
+}
+
+func descriptorToInternal(ocDescriptor *ocmetrics.MetricDescriptor, descriptor data.MetricDescriptor) {
+	if ocDescriptor == nil {
+		return
+	}
+
+	descriptorType := descriptorTypeToInternal(ocDescriptor.Type)
+	if descriptorType == invalidMetricType {
+		return
+	}
+
+	descriptor.SetType(descriptorType)
+	descriptor.SetDescription(ocDescriptor.GetDescription())
+	descriptor.SetName(ocDescriptor.GetName())
+	descriptor.SetUnit(ocDescriptor.GetUnit())
+}
+
+func descriptorTypeToInternal(t ocmetrics.MetricDescriptor_Type) data.MetricType {
+	switch t {
+	case ocmetrics.MetricDescriptor_UNSPECIFIED:
+		return data.MetricTypeUnspecified
+	case ocmetrics.MetricDescriptor_GAUGE_INT64:
+		return data.MetricTypeGaugeInt64
+	case ocmetrics.MetricDescriptor_GAUGE_DOUBLE:
+		return data.MetricTypeGaugeDouble
+	case ocmetrics.MetricDescriptor_GAUGE_DISTRIBUTION:
+		return data.MetricTypeGaugeHistogram
+	case ocmetrics.MetricDescriptor_CUMULATIVE_INT64:
+		return data.MetricTypeCounterInt64
+	case ocmetrics.MetricDescriptor_CUMULATIVE_DOUBLE:
+		return data.MetricTypeCounterDouble
+	case ocmetrics.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		return data.MetricTypeCumulativeHistogram
+	case ocmetrics.MetricDescriptor_SUMMARY:
+		return data.MetricTypeSummary
+	default:
+		return invalidMetricType
+	}
+}
+
+// getInternalLabels scans OC metric labels and returns OC timeseries idx -> labels map.
+func getInternalLabels(ocMetric *ocmetrics.Metric) []data.StringMap {
+	ocLabelsKeys := ocMetric.GetMetricDescriptor().GetLabelKeys()
+	labelsCount := len(ocLabelsKeys)
+	timeseriesCount := len(ocMetric.GetTimeseries())
+	if timeseriesCount == 0 {
+		return nil
+	}
+
+	labelsByTimeseriesIdx := make([]data.StringMap, timeseriesCount)
+
+	// Scan timeseries and fill the OC timeseries idx -> internal labels map
+	for i := 0; i < timeseriesCount; i++ {
+		ts := ocMetric.GetTimeseries()[i]
+		labelsByTimeseriesIdx[i] = data.NewEmptyStringMap(labelsCount)
+		for l := 0; l < labelsCount; l++ {
+			ocLabelValues := ts.GetLabelValues()
+			if ocLabelValues == nil || !ocLabelValues[l].GetHasValue() {
+				continue
+			}
+			labelsByTimeseriesIdx[i].Insert(ocLabelsKeys[l].Key, ocLabelValues[l].Value)
+		}
+	}
+
+	return labelsByTimeseriesIdx
+}
+
+// setDataPoints converts OC timeseries to internal datapoints based on metric type
+func setDataPoints(
+	ocMetric *ocmetrics.Metric,
+	metric data.Metric,
+	labels []data.StringMap,
+) {
+	ocPointsCount := getPointsCount(ocMetric)
+	for i, timeseries := range ocMetric.GetTimeseries() {
+		startTimestamp := internal.TimestampToUnixNano(timeseries.GetStartTimestamp())
+		for _, point := range timeseries.GetPoints() {
+			pointTimestamp := internal.TimestampToUnixNano(point.GetTimestamp())
+			switch point.Value.(type) {
+			case *ocmetrics.Point_Int64Value:
+				if metric.Int64DataPoints().Len() == 0 {
```

Comment:
Please call `metric.Int64DataPoints()` once. Alternatively I would be interested to see a confirmation that Go compiler is smart enough to optimize it and execute once (I am not very hopeful, but happy to be proved wrong).